File: examples/regression.cs
using System;

using nn;

internal unsafe class regression {

    static string pretty_logits(float[] logits) {
        fixed (float* logits0 = logits) {
            return pretty_logits(logits0, (uint)logits.Length);
        }
    }
    static string pretty_logits(float* logits0, uint cc) {
        string row = "[";
        for (int j = 0; j < Math.Min(cc, 7); j++) {
            row += $"{(Math.Round(logits0[j], 4))}";
            if (j == cc - 1)
                row += "";
            else if (j > 0)
                row += ", ";
            else
                row += ", ";
        }
        row += "]";
        return row;
    }

    const int POLY_DEGREE = 4;

    static string poly_desc(Tensor W, Tensor b) {
        string result = "y = ";
        for (int i = 0; i < W.numel(); i++) {
            result += $"+{W.data[i]:f2} x^{i + 1} ";
        }
        result += $"+{b.data[0]:f2}";
        return result;
    }

    static nn.Tensor W_target = new nn.Tensor(POLY_DEGREE, requires_grad: false);
    static nn.Tensor b_target = new nn.Tensor(1, requires_grad: false);

    static (float[], float[]) get_batch(nn.rand.IRNG g, int batch_size = 32) {
        float[] random = new float[batch_size];
        nn.rand.normal_(random, g);
        var x = new float[batch_size * POLY_DEGREE];
        for (int b = 0; b < batch_size; b++) {
            for (int p = 0; p < POLY_DEGREE; p++) {
                x[b * POLY_DEGREE + p] = (float)Math.Pow(random[b], p + 1);
            }
        }
        var y = new float[batch_size];
        for (int b = 0; b < batch_size; b++) {
            y[b] = 0;
            for (int p = 0; p < POLY_DEGREE; p++) {
                y[b] += W_target.data[p] * x[b * POLY_DEGREE + p];
            }
            y[b] += b_target.data[0];
        }
        return (x, y);
    }

    static void Main() {
        var g = new nn.rand.mt19937(137);

        Console.WriteLine(g.randint32());

        nn.rand.uniform_(W_target.data, W_target.numel(), g);
        nn.rand.uniform_(b_target.data, b_target.numel(), g);

        Console.WriteLine(g.randint32());

        Console.WriteLine($"W_target: {pretty_logits(W_target.data, W_target.numel())}");
        Console.WriteLine($"b_target: {pretty_logits(b_target.data, b_target.numel())}");

        var fc = new nn.Linear<F.MatMulV>(POLY_DEGREE, 1);

        nn.rand.kaiming_uniform_(
            fc._Weight.data,
            fc._Weight.numel(),
            g,
            POLY_DEGREE,
            (float)Math.Sqrt(5));

        nn.rand.uniform_(
            fc._Bias.data,
            fc._Bias.numel(),
            g,
            -(float)(1.0/Math.Sqrt(POLY_DEGREE)),
            (float)(1.0 / Math.Sqrt(POLY_DEGREE)));

        Console.WriteLine($"fc.weight: {pretty_logits(fc._Weight.data, fc._Weight.numel())}");
        Console.WriteLine($"fc.bias: {pretty_logits(fc._Bias.data, fc._Bias.numel())}");

        Console.WriteLine(g.randint32());

        double loss = 0; int step;

        for (step = 0; step < 10000; step++) {
            var batch = get_batch(g);

            if (step % 1000 == 0)
                Console.WriteLine($"batch_x: {pretty_logits(batch.Item1)}");

            fc._Weight.zero_grad();
            fc._Bias.zero_grad();

            var input = new Tensor(batch.Item1);

            var logits = fc.forward(input);

            input.Dispose();

            loss = F.mse_loss(logits, batch.Item2);

            fc.backward(logits);

            foreach (var param in fc.parameters()) {
                for (int n = 0; n < param.numel(); n++) {
                    param.data[n] += param.grad[n] * 0.001f;
                }
            }

            if (step % 1000 == 0)
                Console.WriteLine($"step # {step}, loss = {loss}");
        }

        Console.WriteLine($"Loss: {loss:f6} after {step - 1} steps");
        Console.WriteLine($"==> Learned function:\t" + poly_desc(fc._Weight, fc._Bias));
        Console.WriteLine($"==> Actual function:\t" + poly_desc(W_target, b_target));
        Console.ReadKey();
    }
}

File: examples/regression.py
# See https://github.com/pytorch/examples/blob/main/regression/main.py

import torch
import torch.nn.functional as F

def pretty_logits(logits):
    logits0 = logits.view(-1)
    row = "["
    for j in range(min(len(logits0), 7)):
        row += f"{(round(logits0[j].item(), 4)):.4f}"
        if j == len(logits0) - 1:
            row += ""
        elif j > 0 and ((j + 1) % logits.shape[0]) == 0:
            row += ", "
        else:
            row += ", "
            
    row += "]"
    return row

torch.manual_seed(137)

print(torch.randint(0, 0xffffffff, [1]).item())

POLY_DEGREE = 4

W_target = torch.zeros(POLY_DEGREE, 1)
b_target = torch.zeros(1)

W_target.uniform_(0, 1)
b_target.uniform_(0, 1)

print(torch.randint(0, 0xffffffff, [1]).item())

print("W_target: ", pretty_logits(W_target))
print("b_target: ", pretty_logits(b_target))

model = torch.nn.Linear(W_target.size(0), 1)

print("fc.weight: ", pretty_logits(model.weight))
print("fc.bias: ", pretty_logits(model.bias))

print(torch.randint(0, 0xffffffff, [1]).item())

def make_features(x):
    """Builds features i.e. a matrix with columns [x, x^2, x^3, x^4]."""
    x = x.unsqueeze(1)
    return torch.cat([x ** i for i in range(1, POLY_DEGREE+1)], 1)

def f(x):
    """Approximated function."""
    return x.mm(W_target) + b_target.item()

def poly_desc(W, b):
    """Creates a string description of a polynomial."""
    result = 'y = '
    for i, w in enumerate(W):
        result += '{:+.2f} x^{} '.format(w, i + 1)
    result += '{:+.2f}'.format(b[0])
    return result


def get_batch(batch_size=32):
    """Builds a batch i.e. (x, f(x)) pair."""
    random = torch.randn(batch_size)
    x = make_features(random)
    y = f(x)
    return x, y


for step in range(0, 10000):
    batch_x, batch_y = get_batch()
    
    if (step % 1000 == 0):
        print("batch_x: ", pretty_logits(batch_x))

    model.zero_grad()

    output = F.mse_loss(model(batch_x), batch_y)
    loss = output.item()

    output.backward()

    for param in model.parameters():
        param.data.add_(-0.001 * param.grad)

    if (step % 1000 == 0):
        print(f"step # {step}, loss = {loss}")

print('Loss: {:.6f} after {} steps'.format(loss, step))
print('==> Learned function:\t' + poly_desc(model.weight.view(-1), model.bias))
print('==> Actual function:\t' + poly_desc(W_target.view(-1), b_target))